interfaces/builtin/opengl.go

```diff
@@ -45,6 +45,11 @@ const openglConnectedPlugAppArmor = `
   /var/lib/snapd/hostfs/{,usr/}lib{,32,64,x32}/{,@{multiarch}/}lib{GL,EGL}*nvidia.so{,.*} rm,
   /var/lib/snapd/hostfs/{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libGLdispatch.so{,.*} rm,
 
+  # Support reading the Vulkan ICD files
+  /var/lib/snapd/lib/vulkan/ r,
+  /var/lib/snapd/lib/vulkan/** r,
+  /var/lib/snapd/hostfs/usr/share/vulkan/icd.d/10_nvidia*.json r,
```

Comment:
`/var/lib/snapd/hostfs/usr/share/vulkan/icd.d/*.json r,` should catch ICDs from all drivers.

Reply:
And we don't want all drivers, that's the point. We **only** want NVIDIA drivers exposed - not having mesa vulkan exposed :)

Reply:
TLDR a runtime should provide its own mesa and vulkan (`libvulkan.so.1`) like the `solus-runtime-gaming` does, but we accept host ICDs from vulkan to allow the NVIDIA vulkan stuff to work properly.

Reply:
Maybe I'm missing something here, but we're already allowing to use the host's GL implementation. I guess the same would be fitting for Vulkan. Meaning, we should probably allow to access all ICDs. otherwise the client would end up having access to a single implementation only.

If you'd want to use a specific ICD then there's VULKAN_ICD_FILENAMES environment library.

Reply:
No we don't - we only allow the host GL **library** for the NVIDIA library. Allowing mesa to cross-contaminate would be utterly catastrophic. Likewise exposing the invalid ICDs for Vulkan which have the potential to clash with the internal vulkan icds provided by the snaps **own mesa build** will cause things to break.

`solus-runtime-gaming` already has its own mesa, and default vulkan icds for the open source vulkan drivers. We glob this vulkan directory to add the host NVIDIA ICDs as thats the only host GL library we'd ever share. FWIW I'm fully aware of `VULKAN_ICD_FILENAMES`, please read the actual commit messages where I documented that.

Reply:
As an example, the guest may have open source vulkan specifying:
```
    "ICD": {
        "library_path": "/usr/lib64/libvulkan_intel.so",
        "api_version": "1.0.54"
    }
```

And the host could have:

```
    "ICD": {
        "library_path": "/usr/lib64/libvulkan_intel.so",
        "api_version": "1.0.65"
    }
```

By allowing those to contaminate the child snap environment there is every risk of reading the **wrong** ICD file first and `libvulkan` believing that the internally provided driver is capable of a different vulkan version that it actually can, as the implementation that is used does not come from outside the snaps, they come from inside the snaps. Note, for full context, `solus-runtime-gaming` is a **base** snap and cannot be contaminated in this fashion. (Nor should any snaps if you intend confinement to to be the default approach.)

Reply:
LSI reference: https://github.com/solus-project/linux-steam-integration/blob/master/src/shim/shim.c#L150

Reply:
Makes sense now. Thanks for taking the time to explain.

Reply:
Its ugly ugly evil stuff and I wish we didn't have to support it, but there it is :/ FWIW I'll be following up with a highly similar change at some point which will expose the OpenCL ICD definitions (primarily to have CUDA working) - but this will require me to work on a patch to send to Khronos to have something like a `OCL_ICD_FILENAMES` variable too. We'd want this for stuff like Blender+NVCC combinations, etc. Having the context here now though will help when it comes to implementing that next step. :) It'd be down to the internal runtime to provide something like `beignet` as the open source solution.

Reply:
I remember it to be a bit of a hassle with OpenCL. If all else fails you can always try mounting tmpfs over the vendors directory and expose the just file that you need. Ugly as hell, but you wouldn't need to through back and forth with the Khronos group :)

Reply:
This is true, but if necessary we can carry the relevant patch in the runtime, as the ocl-icd package is only tiny for the sake of the patch to actually support it